### PR TITLE
Calibration tour article

### PR DESCRIPTION
**Pull request by topepo**
This will be linked in the Rd files in the next release of probably. Requires the current devel version of probably. Once editing is done, I'll bump the version of probably and re-run before merging.

### CODE REVIEW

File: content/learn/models/calibration/index.Rmarkdown [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+pkgs <- c("tidymodels", "probably", "discrim", "klaR")
+```
+
+
+`r req_pkgs(pkgs)` The probably package should be version 1.0.0 or greater. 
+
+This article demonstrates how to improve an existing model to make its predictions better. A model is well-calibrated if its probability estimate is consistent with the rate that the event occurs "in the wild." In practice though, it can be difficult to validate this definition directly. 
+
+For more details:
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: Should this go at the end?

File: content/learn/models/calibration/index.Rmarkdown [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+library(tidymodels)
+library(probably)
+library(discrim)
+
+tidymodels_prefer()
+theme_set(theme_bw())
+options(pillar.advice = FALSE, pillar.min_title_chars = Inf)
+```
+
+We'll use an old example for illustration.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: I don't think this sentence is needed as a bridge and calling the example old just makes me wonder about that aspect when I had not before 😛 
```suggestion
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+data(cells)
+cells$case <- NULL
+
+dim(cells)
+cells %>% count(class)
+```
+
+There is a class imbalance but that will not affect our work here. 
+
+Let's make a 75%/25% split of the data into training and testing using `initial_split()`. We'll also create a set of 10-fold cross-validation indices for model resampling.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
Let's make a 75% to 25% split of the data into training and testing using `initial_split()`. We'll also create a set of 10-fold cross-validation indices for model resampling. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+cells_te <- testing(split)
+
+cells_rs <- vfold_cv(cells_tr, strata = class)
+```
+
+Now that there are data to be modeled, let's get to it!
+
+## A naive Bayes model
+
+We'll show the utility of calibration tools by using a model that, in this instance, is likely to produce a poorly calibrated model. The naive Bayes classifier is a well established model that assumes that the predictors are statistically _independent_ of one another (to simplify the calculations).  While that is certainly not the case for these data, the model can be effective at discriminating between the classes. Unfortunately, when there are many predictors in the model, it has a tendency to produce class probability distributions that are pathological. The predictions tend to gravitate to values neat zero or one, producing distributions that are "U"-shaped ([Kuhn and Johnson, 2013](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=%22Applied+Predictive+Modeling%22&btnG=)).
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
We'll show the utility of calibration tools by using a type of model that, in this instance, is likely to produce a poorly calibrated model. The naive Bayes classifier is a well-established model that assumes that the predictors are statistically _independent_ of one another (to simplify the calculations).  While that is certainly not the case for these data, the model can be effective at discriminating between the classes. Unfortunately, when there are many predictors in the model, it has a tendency to produce class probability distributions that are pathological. The predictions tend to gravitate to values near zero or one, producing distributions that are "U"-shaped ([Kuhn and Johnson, 2013](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=%22Applied+Predictive+Modeling%22&btnG=)). 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+collect_metrics(bayes_res)
+```
+
+The ROC score is impressive! However, the Brier value indicates that the probability values, while discriminating well, are not very realistic. A value of 0.25 is the "bad model" threshold when there are two classes (a value of zero being the best possible result). 
+
+### But is it calibrated? 
+
+Spoilers: no. It is not. 
+
+The first clue is the extreme U-shaped distribution of the probability scores (facetted by the true class value):
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: extremely?

File: content/learn/models/calibration/index.Rmarkdown [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+  ggplot(aes(.pred_PS)) +
+  geom_histogram(col = "white", bins = 40) +
+  facet_wrap(~ class, ncol = 1) +
+  geom_rug(col = "blue", alpha = 1 / 2) + 
+  labs(x = "Probability Estimate of PS")
+```
+
+There are almost no cells with moderate probability estimates. Furthermore, when the model is incorrect, it is "confidently incorrect". 
+
+The probably package has tools for visualizing and correcting model with poor calibration properties.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
The probably package has tools for visualizing and correcting models with poor calibration properties. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+  facet_wrap(~ class, ncol = 1) +
+  geom_rug(col = "blue", alpha = 1 / 2) + 
+  labs(x = "Probability Estimate of PS")
+```
+
+There are almost no cells with moderate probability estimates. Furthermore, when the model is incorrect, it is "confidently incorrect". 
+
+The probably package has tools for visualizing and correcting model with poor calibration properties. 
+
+The most common plot is to break the predictions into about ten equally sized buckets and compute the actual event rate within each. For example, if a bin captures the samples predicted to be poorly segmented with probabilities between 20% and 30%, we should expect about a 25% event rate within that partition. Here's a plot with ten bins:
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: maybe say that the 25% is the bin midpoint and that's why we are expecting this event rate?

File: content/learn/models/calibration/index.Rmarkdown [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+#| label: break-windowed
+#| out.width: 60%
+#| fig.width: 5
+#| fig.height: 5
+#| fig.align: center
+
+cal_plot_windowed(bayes_res, step_size = 0.025)
+```
+
+Bad. Still bad.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown:173]: 😂

File: content/learn/models/calibration/index.Rmarkdown [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+#| fig.width: 5
+#| fig.height: 5
+#| fig.align: center
+
+cal_plot_windowed(bayes_res, step_size = 0.025)
+```
+
+Bad. Still bad. 
+
+Finally, for two class outcomes, we can fit a logistic regression model use a generalized additive model and examine the trend.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: something went wrong in this sentence :D

File: content/learn/models/calibration/index.Rmarkdown [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+#| fig.align: center
+
+cal_plot_logistic(bayes_res)
+```
+
+Ooof. 
+
+## Remediation
+
+The good news is that we can do something about this. There are tools to fix the probability estimates so that they have better properties.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: Can you be more specific about in which sense we are fixing them? (i.e. which properties are we aiming to improve, and maybe also which we are not aiming to improve if that makes sense and can be stated concisely)

File: content/learn/models/calibration/index.Rmarkdown [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+cal_plot_logistic(bayes_res)
+```
+
+Ooof. 
+
+## Remediation
+
+The good news is that we can do something about this. There are tools to fix the probability estimates so that they have better properties. 
+
+The most common approach is the fit a logistic regression model to the data (with the probability estimates as the predictor). The probability predictions from this model is then used as the calibrated estimate. By default, a generalized additive model is used for this fit, but the `smooth = FALSE` argument can use simple linear effects.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: The central cue to how this actually works comes so very casually in parenthesis here. While the tone is generally casual, I think it's worth expanding on the central idea here in a few more words. Maybe along the lines of that we are fitting a second (or meta) model using the outcome of the first (i.e., those probability estimates) to balance out the shortcomings of the initial model? Is it worth referencing stacking here? This feels like 1-model-stacking to me?

File: content/learn/models/calibration/index.Rmarkdown [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+logit_val <- cal_validate_logistic(bayes_res, metrics = cls_met, save_pred = TRUE)
+collect_metrics(logit_val)
+
+collect_predictions(logit_val) %>%
+  filter(.type == "calibrated") %>%
+  cal_plot_windowed(truth = class, estimate = .pred_PS, step_size = 0.025) +
+  ggtitle("Logistic calibration via GAM")
+```
+
+That's a lot better but it is problematic that the calibrated predictions to not reach zero or one.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
That's a lot better but it is problematic that the calibrated predictions do not reach zero or one. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+
+Ooof. 
+
+## Remediation
+
+The good news is that we can do something about this. There are tools to fix the probability estimates so that they have better properties. 
+
+The most common approach is the fit a logistic regression model to the data (with the probability estimates as the predictor). The probability predictions from this model is then used as the calibrated estimate. By default, a generalized additive model is used for this fit, but the `smooth = FALSE` argument can use simple linear effects. 
+
+How do we know if this works? There are a set of `validate` functions that can use holdout data to resample the model with and without the calibration tool of choice. Since we already resampled the model, we'll use those results to estimate 10 more logistic regressions and use the out-of-sample data to estimate performance.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
How do we know if this works? There are a set of `cal_validate_*()` functions that can use holdout data to resample the model with and without the calibration tool of choice. Since we already resampled the model, we'll use those results to estimate 10 more logistic regressions and use the out-of-sample data to estimate performance. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+collect_predictions(logit_val) %>%
+  filter(.type == "calibrated") %>%
+  cal_plot_windowed(truth = class, estimate = .pred_PS, step_size = 0.025) +
+  ggtitle("Logistic calibration via GAM")
+```
+
+That's a lot better but it is problematic that the calibrated predictions to not reach zero or one. 
+
+A different approach is to use isotonic regression. This method can result in very few unique probability estimates. probably has a version of isotonic regression that resamples the process to produce more unique probabilities:
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
A different approach is to use isotonic regression. This method can result in very few unique probability estimates. The probably package has a version of isotonic regression that resamples the process to produce more unique probabilities: 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+                                      save_pred = TRUE, times = 25)
+collect_metrics(iso_val)
+
+collect_predictions(iso_val) %>%
+  filter(.type == "calibrated") %>%
+  cal_plot_windowed(truth = class, estimate = .pred_PS, step_size = 0.025) +
+  ggtitle("Isotonic regression calibration")
+```
+
+Much better. There is a slight bias since the estimated points are consistently above the identity line on the 45 degree angle.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
Much better. However, there is a slight bias since the estimated points are consistently above the identity line on the 45-degree angle. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+
+Beta calibration appears to have the best results. We'll save a model that is trained using all of the out-of-sample predictions from the original naive Bayes resampling results. The `cell_cal` object can be used to enact the calibration for new predictions. 
+
+We can also fit the final naive Bayes model to predict the test set: 
+
+```{r}
+#| label: finalize
+#| warning: false
+
+cell_cal <- cal_estimate_beta(bayes_res)
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown:269]: oh, this is a `cal_estimate_*()` function now! It might be worth adding a sentence on the "validate" vs the "estimate" functions so that people are aware of them as two groups of functions

File: content/learn/models/calibration/index.Rmarkdown [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+collect_predictions(beta_val) %>%
+  filter(.type == "calibrated") %>%
+  cal_plot_windowed(truth = class, estimate = .pred_PS, step_size = 0.025) +
+  ggtitle("Beta calibration")
+```
+
+Also a big improvement but it does poorly at the lower end of the scale. 
+
+Beta calibration appears to have the best results. We'll save a model that is trained using all of the out-of-sample predictions from the original naive Bayes resampling results. The `cell_cal` object can be used to enact the calibration for new predictions.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: It be nice to show the code making `cell_cal` before referencing it in the text

File: content/learn/models/calibration/index.Rmarkdown [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+
+```{r}
+#| label: uncalibrated
+#| warning: false
+
+cell_test_pred <- augment(bayes_fit, new_data = cells_te)
+cell_test_pred %>% cls_met(class, .pred_PS)
+```
+
+These metric estimates are very consistent with the resasmpled performance estimates.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
These metric estimates are very consistent with the resampled performance estimates. 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+cell_test_cal_pred %>% cls_met(class, .pred_PS)
+cell_test_cal_pred %>%
+  cal_plot_windowed(truth = class, estimate = .pred_PS, step_size = 0.025)
+```
+
+Much better. The test set results also agree with the results from `cal_validate_beta().` 
+
+## Other model types
+
+probably can also calibrate classification models with more than two outcome levels. The functions `cal_*_multinomial()` uses a multinomial model in the same spirit as the logistic regression model. Also, isotonic and Beta calibration can also be used via a "one versus all" approach that builds a set of binary calibrators and normalizes their results at the end (to ensure that they add to one).
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]:
```suggestion
probably can also calibrate classification models with more than two outcome levels. The functions `cal_*_multinomial()` use a multinomial model in the same spirit as the logistic regression model. Isotonic and Beta calibration can also be used via a "one versus all" approach that builds a set of binary calibrators and normalizes their results at the end (to ensure that they add to one). 
```

File: content/learn/models/calibration/index.Rmarkdown [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+Much better. The test set results also agree with the results from `cal_validate_beta().` 
+
+## Other model types
+
+probably can also calibrate classification models with more than two outcome levels. The functions `cal_*_multinomial()` uses a multinomial model in the same spirit as the logistic regression model. Also, isotonic and Beta calibration can also be used via a "one versus all" approach that builds a set of binary calibrators and normalizes their results at the end (to ensure that they add to one). 
+
+For regression models, there is `cal_plot_regression()` and `cal_*_linear()`. The latter uses `lm()` or `mgcv::gam()` to create a calibrator object. 
+
+## Future plans
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: Should this go here? I typically think of tidymodels.org articles as reference posts rather than updates. Because of this, that section feels slightly out of place here (as opposed to in a blog post).

File: content/learn/models/calibration/index.Rmarkdown [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+
+source(here::here("content/learn/common.R"))
+
+pkgs <- c("tidymodels", "probably", "discrim", "klaR")
+```
+
+
+`r req_pkgs(pkgs)` The probably package should be version 1.0.0 or greater. 
+
+This article demonstrates how to improve an existing model to make its predictions better. A model is well-calibrated if its probability estimate is consistent with the rate that the event occurs "in the wild." In practice though, it can be difficult to validate this definition directly.
```
Comment (hfrick) [inline content/learn/models/calibration/index.Rmarkdown]: That last sentence on validation is kinda just dangling there for me. I would probably remove it and just run with the core definition of which problem we are addressing in this post. Or elaborate a bit more in general.

File: content/learn/models/calibration/index.Rmarkdown (hunk 20 above)
Comment (hfrick) [reply content/learn/models/calibration/index.Rmarkdown]: Actually, I think this should go at the start. Not as "workflows can't do this yet" but as "what we are doing here is post-processing the predictions"

**Review by hfrick — COMMENTED** [review]
I like it and I think it's an exciting addition to tidymodels! 

I would add a little bit more to what this is and why it's useful. I feel like its importance is not shining through as much as it should here! I would situate it as post-processing (even if it's not yet integrated into workflows):
Since we are trying to improve predictions, I think it would be helpful to establish where this sits in relation to "make your prediction _model_  better" (i.e. feature engineering, hyperparameter tuning, etc) and to clarify when you would go through the trouble of post-processing your predictions.

File: content/learn/models/calibration/index.Rmarkdown (hunk 20 above)
Comment (topepo) [reply content/learn/models/calibration/index.Rmarkdown]: I was going to make small articles for multi-class and regression models. Should I just take these last two sections out?

Comment (hfrick) [reply content/learn/models/calibration/index.Rmarkdown]: I like the "Other model types" section. I think it could stay even when there are more specific articles on them.
